framework/finalizer.go

```diff
@@ -0,0 +1,99 @@
+package framework
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/rest"
+)
+
+const (
+	finalizerName = "operatorkit.giantswarm.io"
+)
+
+type patchSpec struct {
+	Op    string   `json:"op"`
+	Path  string   `json:"path"`
+	Value []string `json:"value"`
+}
+
+func (f *Framework) addFinalizer(obj interface{}) (bool, error) {
+	accessor, err := meta.Accessor(obj)
+	if err != nil {
+		return false, err
+	}
+	if accessor.GetDeletionTimestamp() != nil {
+		return true, nil // object has been marked for deletion, we should ignore it.
+	}
+	if containsFinalizer(accessor.GetFinalizers(), finalizerName) {
+		return false, nil // object already has the finalizer.
+	}
+
+	c, err := rest.RESTClientFor(&rest.Config{})
+	if err != nil {
+		return false, err
+	}
+	patch := fmt.Sprintf(`{"op":"add","value":%q,"path":"/metadata/finalizers/1"}`, finalizerName)
+	res := c.Patch(types.JSONPatchType).
+		AbsPath(accessor.GetSelfLink()).
+		Body(patch).
+		Do()
+	if res.Error() != nil {
+		return false, res.Error()
+	}
+
+	return true, nil
+}
+
+func (f *Framework) removeFinalizer(obj interface{}) error {
+	accessor, err := meta.Accessor(obj)
+	if err != nil {
+		return err
+	}
+	if !containsFinalizer(accessor.GetFinalizers(), finalizerName) {
+		return nil // resource has no finalizer, probably migration.
+	}
+	c, err := rest.RESTClientFor(&rest.Config{})
+	if err != nil {
+		return err
+	}
+
+	patch := patchSpec{
```

Comment:
This is a bit more complicated. JSONPatch does not seem to have a way to delete only one entry of an array without specifying its index. (see http://jsonpatch.com/#remove and https://stackoverflow.com/questions/27439986/what-is-the-json-patch-format-to-remove-an-element-from-an-array#comment48742836_27458369 )

This did not seem safe to me, so instead we take all finalizers, remove our finalizer and then replace the entire array.

Reply:
Sorry, missed this comment in my previous review. Couldn't you find the index to be deleted first and then apply the remove specifying that index?

Reply:
Yes, I thought about this - honestly I am not sure how the index is preserved through all this (e.g. a finalizer gets removed before ours ->shifting the list -> we would remove the wrong finalizer using the index; when replacing we at worst add one finalizer back which just causes another reconciliation). So I felt that it was more robust to replace. Do you know if it's safe to assume that the index stays the same?

Reply:
You can use `metadata.resourceVersion` for that.

@fgimenez comment is legit. The approach taken here is vulnerable to races. An already removed finalizer can be readded.

Reply:
I see the point, I just don't understand how I can supply the `metadata.resourceVersion` in my Patch. By simply taking the index without ensuring the `resourceVersion` we might delete the wrong finalizer (as I described above).

Reply:
Also another thought: Readding a removed finalizer should not work. The object would only be allowed to have finalizers removed. If we accidentally add back a finalizer through the `replace` operation we should receive an error from the REST api.

Reply:
Where can be removed those finalizers? If the only place they can be removed is in this `removeFinalizer` method and you are concerned about concurrent execution you could add a `sync.Mutex` and protect the patch call, WDYT? Or they can be removed by another process?

Reply:
Multiple operators could remove finalizers from the same object concurrently (e.g. multiple operators watching a Pod which had an deletion event)

Reply:
Framework processes events one-by-one. Problem is when finalizer is deleted by some external source.

Reply:
I don't see how to pass resourceVersion either.

But just to be safe you can also test the value: http://jsonpatch.com/#test and pass it together with deletion 

You can see if any op fails, whole patch is aborted: http://jsonpatch.com/#how-it-works

Reply:
You can do it in backoff.Retry to mitigate races.